p2p/src/protocol/connmgr.rs

```diff
@@ -279,36 +293,31 @@ impl<U: Connect + Disconnect + Events + SetTimeout, A: AddressSource> Connection
 
     /// Attempt to maintain a certain number of outbound peers.
     fn maintain_connections<S: peer::Store>(&mut self, addrs: &A) {
-        while self.outbound().count()
+        let current = self.outbound().count()
             + self
                 .peers
                 .values()
                 .filter(|p| matches!(p, Peer::Connecting))
-                .count()
-            < self.config.target_outbound_peers
-        {
-            // Prefer addresses with the preferred services.
-            let result = addrs
-                .sample(self.config.preferred_services)
-                .or_else(|| addrs.sample(self.config.required_services));
-
-            if let Some((addr, source)) = result {
-                // TODO: Support Tor?
-                if let Ok(sockaddr) = addr.socket_addr() {
-                    // TODO: Remove this assertion once address manager no longer cares about
-                    // connections.
-                    debug_assert!(!self.is_connected(&sockaddr));
-
-                    if self.connect::<S>(&sockaddr) {
-                        self.upstream.event(Event::Connecting(sockaddr, source));
-                        break;
-                    }
+                .count();
+        let target = self.config.target_outbound_peers;
+        let addresses = addrs
+            .iter(self.config.preferred_services)
+            .chain(addrs.iter(self.config.required_services))
+            .take(target - current);
+
+        // TODO: The address manager currently may return duplicates if the address book is small
+        // and we are requested more addresses than it has.
+
+        for (addr, source) in addresses {
+            // TODO: Support Tor?
+            if let Ok(sockaddr) = addr.socket_addr() {
+                // TODO: Remove this assertion once address manager no longer cares about
+                // connections.
+                debug_assert!(!self.is_connected(&sockaddr));
```

Comment:
```suggestion
                debug_assert!(!self.is_connected(&sockaddr) || !self.is_connecting(&sockaddr));
```

Reply:
So not sure this is safe actually. The reason we know for sure that the address returned isn't connected, is that the addrmgr keeps track of nodes we connected to. But it doesn't always keep track of peers that are attempted (connecting), only if it's an address it knows.